pkg/platform/kube/functionres/lazy.go

```diff
@@ -634,10 +642,90 @@ func (lc *lazyClient) createOrUpdateDeployment(functionLabels labels.Set,
 	return resource.(*apps_v1beta1.Deployment), err
 }
 
+func (lc *lazyClient) canUpdateDeploymentStrategy(deployment *apps_v1beta1.Deployment, function *nuclioio.NuclioFunction) (bool, error) {
+	deploymentAugmentedConfigs, err := lc.getDeploymentAugmentedConfigs(function)
+	if err != nil {
+		return false, errors.Wrap(err, "Failed to get deployment augmented configs")
+	}
+
+	// check if user didnt provide a deployment strategy
+	for _, augmentedConfig := range deploymentAugmentedConfigs {
+		if augmentedConfig.Kubernetes.Deployment.Spec.Strategy.Type != "" ||
+			augmentedConfig.Kubernetes.Deployment.Spec.Strategy.RollingUpdate != nil {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (lc *lazyClient) updateDeploymentStrategy(deployment *apps_v1beta1.Deployment, function *nuclioio.NuclioFunction) (*apps_v1beta1.Deployment, error) {
+	var jsonPatchMapper []map[string]string
+	var nextDeploymentStrategyType apps_v1beta1.DeploymentStrategyType
+
+	// check user didn't provide any deployment strategy specifics
+	if canUpdateDeploymentStrategy, err := lc.canUpdateDeploymentStrategy(deployment, function); err != nil {
+		return nil, errors.Wrap(err, "Failed to get deployment augmented configs")
+	} else if !canUpdateDeploymentStrategy {
+		return deployment, nil
+	}
+
+	// Since k8s (ATM) does not support rolling update for GPU
+	// redeploying a Nuclio function will get stuck if no GPU is available
+	// to overcome it, we simply change the update strategy to recreate
+	// so k8s will kill the existing pod\function and create the new one
+	if gpuResource, ok := function.Spec.Resources.Limits[nvidiaGpuResourceName]; ok {
+
+		// user specifically asked for 0 gpus, retain the rolling update
+		if gpuResource.IsZero() {
+			nextDeploymentStrategyType = apps_v1beta1.RollingUpdateDeploymentStrategyType
+		} else {
+
+			// requested gpus, change to recreate
+			nextDeploymentStrategyType = apps_v1beta1.RecreateDeploymentStrategyType
+		}
+	} else {
+
+		// no gpu resources, set to rollingUpdate
+		nextDeploymentStrategyType = apps_v1beta1.RollingUpdateDeploymentStrategyType
+	}
```

Comment:
Redundant with many elses.
Can be written as
`newDeploymentStrategyType = lc.resolveDefaultDeploymentStrategy(function)`
This method's logic is exactly the old `populateDefaultDeploymentStrategy` except it shouldn't get a pointer to `strategy` but just return the resolved default one. it's logic is the same as here.

This also means `resolveDefaultDeploymentStrategy` can be called directly from `enrichDeploymentFromPlatformConfiguration` on creation.
In order to keep it from happening on update, you can simply pass a boolean to branch the behavior of `enrichDeploymentFromPlatformConfiguration` which then means you can remove the problematic `updateDeploymentStrategy` which happens right after creation because it will be created correctly like before